src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::sql::optimizer::rule::Rule;
+use crate::sql::optimizer::rule::TransformState;
+use crate::sql::optimizer::RuleID;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::JoinType;
+use crate::sql::plans::Limit;
+use crate::sql::plans::LogicalInnerJoin;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::RelOp;
+use crate::sql::plans::RelOperator;
+
+/// Input:      Limit
+///               |
+///        Left Outer Join
+///             /    \
+///            *      *
+///
+/// Output:
+///        Left Outer Join
+///             /    \
+///           Limit   *
+///           /
+///          *
+pub struct RulePushDownLimitOuterJoin {
+    id: RuleID,
+    pattern: SExpr,
+}
```

Comment:
Is this a legal rule?

For example,
```sql
create table t(a int);
create table t1(a int);
insert into t values(1);
insert into t1 values(1),(1);

select * from t left join t1 on t.a = t1.a limit 1;
 a | a
---+---
 1 | 1
(1 row)

select * from (select * from t limit 1) as t left join t1 on t.a = t1.a;
 a | a
---+---
 1 | 1
 1 | 1
(2 rows)
```
The result is different after this push down.

Reply:
If want to push limit with join,  maybe the rule need modify to this ?

```
/// Input:      Limit
///               |
///        Left Outer Join
///             /    \
///            *      *
///
/// Output:
///             limit
///              |
///        Left Outer Join
///             /    \
///           Limit   *
///           /
///          *
```

Reply:
@TCeason Yes, it's possible to push the limit down to both sides of outer join.

Reply:
Add test in ther: tests/logictest/suites/base/03_dml/03_0010_select_limit_offset